crates/cubecl-wgpu/src/compiler/wgsl/instructions.rs

```diff
@@ -1012,51 +1035,47 @@ fn index(
     rhs: &Variable,
     out: &Variable,
     offset: Option<Variable>,
+    len: Option<&Variable>,
 ) -> core::fmt::Result {
     let is_scalar = match lhs {
         Variable::Local { item, .. } => item.vectorization_factor() == 1,
         Variable::LocalBinding { item, .. } => item.vectorization_factor() == 1,
         _ => false,
     };
 
-    if out.item().elem().is_atomic() {
-        match offset {
-            Some(offset) => writeln!(f, "let {out} = &{lhs}[{rhs} + {offset}];"),
-            None => writeln!(f, "let {out} = &{lhs}[{rhs}];"),
-        }
-    } else if lhs.elem() != out.elem() {
-        let item = out.item();
-        let out = out.fmt_left();
-        match offset {
-            Some(offset) => {
-                let value = lhs
-                    .item()
-                    .fmt_cast_to(item, format!("{lhs}[{rhs}+{offset}]"));
-                writeln!(f, "{out} = {value};")
-            }
-            None => {
-                if is_scalar {
-                    let value = lhs.item().fmt_cast_to(item, format!("{lhs}"));
-                    writeln!(f, "{out} = {value};")
-                } else {
-                    let value = lhs.item().fmt_cast_to(item, format!("{lhs}[{rhs}]"));
-                    writeln!(f, "{out} = {value};")
-                }
-            }
-        }
+    let (mut value, index) = if is_scalar {
+        (format!("{lhs}"), None)
     } else {
-        let out = out.fmt_left();
-        match offset {
-            Some(offset) => writeln!(f, "{out} = {lhs}[{rhs} + {offset}];"),
-            None => {
-                if is_scalar {
-                    writeln!(f, "{out} = {lhs};")
-                } else {
-                    writeln!(f, "{out} = {lhs}[{rhs}];")
-                }
-            }
-        }
+        let index = if let Some(offset) = offset {
+            format!("{rhs}+{offset}")
+        } else {
+            format!("{rhs}")
+        };
+
+        (format!("{lhs}[{index}]"), Some(index))
+    };
+
+    if out.item().elem().is_atomic() {
+        value = format!("atomicLoad(&{value})")
+    };
+
+    if lhs.elem() != out.elem() {
+        value = lhs.item().fmt_cast_to(out.item(), value)
+    };
+
+    if let Some(ind) = index {
+        if let Some(len) = len {
+            // Note: This is technically not 100% allowed. According to the WebGPU specification,
+            // any OOB access is a "dynamic error" which allows "many possible outcomes". In practice,
+            // both wgpu and Dawn handle this by either returning dummy data or clamping the index
+            // to valid bounds. This means it's harmless to use in a select.
+            let out_item = out.item();
+            value = format!("select({out_item}(0), {value}, {ind} < {len})");
```

Comment:
Not sure this is going to work with vectorization.

Reply:
Just with the syntax of vec4(0)? Or more that indexing needs to take vectorization into account? I'm really not sure how vecorization works, if  you could add a test for it that'd be amazing!